[BUG]: CI: aarch64 build jobs run emulated, adding a flaky QEMU setup step and hour-long build times

### Tvheadend Version

v4.3 (stable)

### Build Number

n/a (CI infrastructure; workflows as of master a502d063c)

### Platform

Other.. (Fedora 44)

### Architecture

ARMv8 (aarch64)

### Category

Other..

### Description

On the CI run for PR #2186, the job "Build on bookworm aarch64" failed after 19 seconds:

```
Unable to find image 'tonistiigi/binfmt:latest' locally
docker: Error response from daemon: Head "https://registry-1.docker.io/v2/tonistiigi/binfmt/manifests/latest": received unexpected HTTP status: 502 Bad Gateway
##[error]The process '/home/runner/work/_actions/uraimo/run-on-arch-action/v3/src/run-on-arch.sh' failed with exit code 125
```

Run: https://github.com/tvheadend/tvheadend/actions/runs/30647457143, job: https://github.com/tvheadend/tvheadend/actions/runs/30647457143/job/91213673067

The job died in QEMU emulation setup, before checking out or compiling anything. No change to the tree can cause or prevent this failure; it is a property of how the job is hosted.

The structural picture: the "Build on \<distro\> \<arch\>" matrix in `build-ci.yml` (and the same matrix in `build-cloudsmith.yml`) runs its 10 aarch64 legs under QEMU user-mode emulation on amd64 runners, via `uraimo/run-on-arch-action`. That hosting choice has two costs:

1. **An avoidable failure mode.** The action's setup unconditionally pulls and runs `tonistiigi/binfmt` from Docker Hub (`src/run-on-arch.sh`, `install_deps`), so every leg carries a third-party registry dependency that can kill the job before the build starts, as above.
2. **Emulation speed.** In the run linked above the emulated aarch64 legs took between 29 and 97 minutes; the three slowest jobs of the entire run were all aarch64. These legs set the wall-clock time of every CI run.

The same workflow already demonstrates the alternative. The almalinux jobs run a runner matrix that includes `ubuntu-24.04-arm`, GitHub's free native arm64 runners (`build-ci.yml` line 323), and in that same run the two native aarch64 almalinux builds finished in 1 and 2 minutes. The workloads are not identical, so that is not a clean ratio, but the gap is dominated by emulation.

Suggested minimal fix: select the runner by target architecture in the run-on-arch jobs of both workflows, so the aarch64 legs execute natively while everything else is untouched:

```diff
-    runs-on: ubuntu-latest
+    runs-on: ${{ matrix.arch == 'aarch64' && 'ubuntu-24.04-arm' || 'ubuntu-latest' }}
```

On an arm64 host, Docker runs the action's arm64v8-based distro containers natively, so the builds themselves stop being emulated. Two caveats, stated so the limits of the one-liner are clear:

- `run-on-arch-action` v3 still performs its binfmt setup unconditionally, so the Docker Hub pull (and its failure mode) remains until either the action learns to skip it on a matching host, or the aarch64 legs are given the almalinux treatment (a plain `container:` on the arm runner), which is a slightly larger rework and would remove the dependency entirely.
- The armv6/armv7 legs must stay where they are: GitHub's arm64 runners are Azure Cobalt 100 (Neoverse N2 family) without 32-bit AArch32 execution, so 32-bit containers cannot run natively there, and QEMU user-mode emulation is roughly the same speed on either host.

Deliberately out of scope: the "Build on Raspberry Pi" jobs use `pguyot/arm-runner-action` full-system emulation with pinned CPU models and faked cpuinfo, which looks intentional and is left alone. The `linux/arm64` buildx platform in `container-build.yaml` could also move to a native runner with a manifest merge, but that is a separable change.

The practical gain beyond the hour of wall clock: infrastructure failures on these legs are common enough that unrelated red already needs explaining away in PRs (see https://github.com/tvheadend/tvheadend/pull/2179#issuecomment-4995136080 for a different flavour, apt against EOL archives, which native runners would not fix). Removing the emulation-specific class of failure and shrinking the matrix tail makes a red leg mean something again.

Happy to submit the minimal fix as a PR if that helps. It has a convenient property: `build-ci.yml` is `pull_request`-triggered, and pull_request runs use the workflow file from the PR itself, so the PR's own checks would show the aarch64 legs building natively before anything is merged. The `build-cloudsmith.yml` half can only prove itself after merge, since that workflow triggers on pushes to master (and ignores `.github` paths, so the merge itself would not trigger it either; its first native run would come with the next code push).

### Logfile

https://github.com/tvheadend/tvheadend/actions/runs/30647457143/job/91213673067 (full log; the failure excerpt is quoted in the description)

### Reproduction

1. Trigger the build workflows on any branch or PR.
2. Observe each `aarch64` leg of the "Build on \<distro\> \<arch\>" matrix spending its first step installing QEMU binfmt handlers (a `docker run` of `tonistiigi/binfmt` pulled from Docker Hub) and then building under emulation for 29 to 97 minutes.
3. Intermittently, the Docker Hub pull fails and the leg dies before checkout, as in the linked run.

--- a/.github/workflows/build-ci.yml
+++ b/.github/workflows/build-ci.yml
@@ -81,7 +81,9 @@ jobs:
     # Run steps on a matrix
     strategy:
       matrix:
-       arch: [ armv7, aarch64]
+       # aarch64 is not emulated here: it builds natively on arm64
+       # runners in build-deb-native below.
+       arch: [ armv7 ]
        distro: [ stretch, buster, bullseye, bookworm, ubuntu16.04, ubuntu18.04, ubuntu20.04, ubuntu22.04, ubuntu24.04 ]
        include:
          - arch: armv6
@@ -90,8 +92,6 @@ jobs:
            distro: buster
          - arch: armv6
            distro: bullseye
-         - arch: aarch64
-           distro: ubuntu_latest
 
     steps:
       - uses: actions/checkout@v3
@@ -181,21 +181,53 @@ jobs:
           if-no-files-found: error
 
   build-deb-native:
-    runs-on: ubuntu-latest
+    runs-on: ${{ matrix.runner }}
     needs: vue
     continue-on-error: true
-    name: Build on native ${{ matrix.container[1] }}
+    name: Build on ${{ matrix.container[2] || format('native {0}', matrix.container[1]) }}
 
     strategy:
       matrix:
+        runner: ["ubuntu-latest"]
         container: [["i386/ubuntu:trusty", "i386-ubuntu-trusty"], ["ubuntu:trusty", "ubuntu-trusty"], ["i386/ubuntu:xenial", "i386-ubuntu-xenial"], ["ubuntu:xenial", "ubuntu-xenial"], ["ubuntu:bionic", "ubuntu-bionic"], ["ubuntu:focal", "ubuntu-focal"], ["ubuntu:jammy", "ubuntu-jammy"], ["ubuntu:noble", "ubuntu-noble"], ["ubuntu:plucky", "ubuntu-plucky"], ["ubuntu:questing", "ubuntu-questing"], ["ubuntu:resolute", "ubuntu-resolute"], ["i386/debian:stretch", "i386-debian-stretch"], ["debian:stretch", "debian-stretch"], ["i386/debian:buster", "i386-debian-buster"], ["debian:buster", "debian-buster"], ["i386/debian:bullseye", "i386-debian-bullseye"], ["debian:bullseye", "debian-bullseye"], ["i386/debian:bookworm", "i386-debian-bookworm"], ["debian:bookworm", "debian-bookworm"], ["i386/debian:trixie", "i386-debian-trixie"], ["debian:trixie", "debian-trixie"], ["i386/debian:sid", "i386-debian-sid"], ["debian:sid", "debian-sid"]]
+        # aarch64 legs on GitHub's arm64 runners, replacing the QEMU-emulated
+        # aarch64 builds formerly in build-crosscompile:
+        # uraimo/run-on-arch-action cannot run on arm64 hosts (its QEMU
+        # helper image is amd64-only, uraimo/run-on-arch-action#161).
+        # Each include entry overrides both matrix axes, so it is added as
+        # its own combination instead of cross-multiplying with the
+        # ubuntu-latest legs above.
+        # The third container element is the display name: these legs keep
+        # the job names of the QEMU builds they replace, so the required
+        # status checks on master keep matching by name.
+        include:
+          - runner: ubuntu-24.04-arm
+            container: ["arm64v8/debian:stretch", "arm64-debian-stretch", "stretch aarch64"]
+          - runner: ubuntu-24.04-arm
+            container: ["arm64v8/debian:buster", "arm64-debian-buster", "buster aarch64"]
+          - runner: ubuntu-24.04-arm
+            container: ["arm64v8/debian:bullseye", "arm64-debian-bullseye", "bullseye aarch64"]
+          - runner: ubuntu-24.04-arm
+            container: ["arm64v8/debian:bookworm", "arm64-debian-bookworm", "bookworm aarch64"]
+          - runner: ubuntu-24.04-arm
+            container: ["arm64v8/ubuntu:xenial", "arm64-ubuntu-xenial", "ubuntu16.04 aarch64"]
+          - runner: ubuntu-24.04-arm
+            container: ["arm64v8/ubuntu:bionic", "arm64-ubuntu-bionic", "ubuntu18.04 aarch64"]
+          - runner: ubuntu-24.04-arm
+            container: ["arm64v8/ubuntu:focal", "arm64-ubuntu-focal", "ubuntu20.04 aarch64"]
+          - runner: ubuntu-24.04-arm
+            container: ["arm64v8/ubuntu:jammy", "arm64-ubuntu-jammy", "ubuntu22.04 aarch64"]
+          - runner: ubuntu-24.04-arm
+            container: ["arm64v8/ubuntu:noble", "arm64-ubuntu-noble", "ubuntu24.04 aarch64"]
+          - runner: ubuntu-24.04-arm
+            container: ["arm64v8/ubuntu:latest", "arm64-ubuntu-latest", "ubuntu_latest aarch64"]
     steps:
       - name: Setup container
         run: |
           docker pull ${{ matrix.container[0] }}
           docker run --name build-container -d -e GITHUB_ACTIONS=true -v ${{ github.workspace }}:/workspace ${{ matrix.container[0] }} tail -f /dev/null
       - name: Fix old debian apt
-        if: matrix.container[0] == 'debian:stretch' || matrix.container[0] == 'i386/debian:stretch' || matrix.container[0] == 'debian:buster' || matrix.container[0] == 'i386/debian:buster'
+        if: endsWith(matrix.container[0], 'debian:stretch') || endsWith(matrix.container[0], 'debian:buster')
         env:
           SCRIPT: |
             sed -i 's/deb.debian.org/archive.debian.org/g' /etc/apt/sources.list

--- a/.github/workflows/build-cloudsmith.yml
+++ b/.github/workflows/build-cloudsmith.yml
@@ -81,7 +81,9 @@ jobs:
     # Run steps on a matrix
     strategy:
       matrix:
-       arch: [ armv7, aarch64]
+       # aarch64 is not emulated here: it builds natively on arm64
+       # runners in build-deb-native below.
+       arch: [ armv7 ]
        distro: [ stretch, buster, bullseye, bookworm, ubuntu16.04, ubuntu18.04, ubuntu20.04, ubuntu22.04, ubuntu24.04, ubuntu_latest ]
        include:
          - arch: armv6
@@ -189,22 +191,54 @@ jobs:
           if-no-files-found: error
 
   build-deb-native:
-    runs-on: ubuntu-latest
+    runs-on: ${{ matrix.runner }}
     needs: vue
     continue-on-error: true
-    name: Build on native ${{ matrix.container[1] }}
+    name: Build on ${{ matrix.container[2] || format('native {0}', matrix.container[1]) }}
     env:
       PCLOUD_TOKEN: ${{ secrets.PCLOUD_TOKEN }}
     strategy:
       matrix:
+        runner: ["ubuntu-latest"]
         container: [["i386/ubuntu:trusty", "i386-ubuntu-trusty"], ["ubuntu:trusty", "ubuntu-trusty"], ["i386/ubuntu:xenial", "i386-ubuntu-xenial"], ["ubuntu:xenial", "ubuntu-xenial"], ["ubuntu:bionic", "ubuntu-bionic"], ["ubuntu:focal", "ubuntu-focal"], ["ubuntu:jammy", "ubuntu-jammy"], ["ubuntu:noble", "ubuntu-noble"], ["ubuntu:plucky", "ubuntu-plucky"], ["ubuntu:questing", "ubuntu-questing"], ["ubuntu:resolute", "ubuntu-resolute"], ["i386/debian:stretch", "i386-debian-stretch"], ["debian:stretch", "debian-stretch"], ["i386/debian:buster", "i386-debian-buster"], ["debian:buster", "debian-buster"], ["i386/debian:bullseye", "i386-debian-bullseye"], ["debian:bullseye", "debian-bullseye"], ["i386/debian:bookworm", "i386-debian-bookworm"], ["debian:bookworm", "debian-bookworm"], ["i386/debian:trixie", "i386-debian-trixie"], ["debian:trixie", "debian-trixie"], ["i386/debian:sid", "i386-debian-sid"], ["debian:sid", "debian-sid"]]
+        # aarch64 legs on GitHub's arm64 runners, replacing the QEMU-emulated
+        # aarch64 builds formerly in build-crosscompile:
+        # uraimo/run-on-arch-action cannot run on arm64 hosts (its QEMU
+        # helper image is amd64-only, uraimo/run-on-arch-action#161).
+        # Each include entry overrides both matrix axes, so it is added as
+        # its own combination instead of cross-multiplying with the
+        # ubuntu-latest legs above.
+        # The third container element is the display name: these legs keep
+        # the job names of the QEMU builds they replace, so the required
+        # status checks on master keep matching by name.
+        include:
+          - runner: ubuntu-24.04-arm
+            container: ["arm64v8/debian:stretch", "arm64-debian-stretch", "stretch aarch64"]
+          - runner: ubuntu-24.04-arm
+            container: ["arm64v8/debian:buster", "arm64-debian-buster", "buster aarch64"]
+          - runner: ubuntu-24.04-arm
+            container: ["arm64v8/debian:bullseye", "arm64-debian-bullseye", "bullseye aarch64"]
+          - runner: ubuntu-24.04-arm
+            container: ["arm64v8/debian:bookworm", "arm64-debian-bookworm", "bookworm aarch64"]
+          - runner: ubuntu-24.04-arm
+            container: ["arm64v8/ubuntu:xenial", "arm64-ubuntu-xenial", "ubuntu16.04 aarch64"]
+          - runner: ubuntu-24.04-arm
+            container: ["arm64v8/ubuntu:bionic", "arm64-ubuntu-bionic", "ubuntu18.04 aarch64"]
+          - runner: ubuntu-24.04-arm
+            container: ["arm64v8/ubuntu:focal", "arm64-ubuntu-focal", "ubuntu20.04 aarch64"]
+          - runner: ubuntu-24.04-arm
+            container: ["arm64v8/ubuntu:jammy", "arm64-ubuntu-jammy", "ubuntu22.04 aarch64"]
+          - runner: ubuntu-24.04-arm
+            container: ["arm64v8/ubuntu:noble", "arm64-ubuntu-noble", "ubuntu24.04 aarch64"]
+          - runner: ubuntu-24.04-arm
+            container: ["arm64v8/ubuntu:latest", "arm64-ubuntu-latest", "ubuntu_latest aarch64"]
     steps:
       - name: Setup container
         run: |
           docker pull ${{ matrix.container[0] }}
           docker run --name build-container -d -e GITHUB_ACTIONS=true -v ${{ github.workspace }}:/workspace ${{ matrix.container[0] }} tail -f /dev/null
       - name: Fix old debian apt
-        if: matrix.container[0] == 'debian:stretch' || matrix.container[0] == 'i386/debian:stretch' || matrix.container[0] == 'debian:buster' || matrix.container[0] == 'i386/debian:buster'
+        if: endsWith(matrix.container[0], 'debian:stretch') || endsWith(matrix.container[0], 'debian:buster')
         env:
           SCRIPT: |
             sed -i 's/deb.debian.org/archive.debian.org/g' /etc/apt/sources.list

